oddt/fingerprints.py

```diff
@@ -324,34 +330,46 @@ def _ECFP_atom_hash(mol, idx, depth=2, use_pharm_features=False,
     """
     atom_env = [[idx]]
     for r in range(1, depth + 1):
-        prev_atom_env = atom_env[r - 1]
-        if r > 2:  # prune visited atoms
-            prev_atom_env = prev_atom_env[len(atom_env[r - 2]):]
         tmp = []
-        for atom_idx in prev_atom_env:
-            # Toolkit independent version (slower 30%)
-            # for neighbor in mol.atoms[atom_idx].neighbors:
-            #     if neighbor.atomicnum == 1:
-            #         continue
-            #     n_idx = neighbor.idx0
-            #     if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-            #         tmp.append(n_idx)
-            if hasattr(mol, 'OBMol'):
-                for neighbor in oddt.toolkits.ob.OBAtomAtomIter(mol.OBMol.GetAtom(atom_idx + 1)):
-                    if neighbor.GetAtomicNum() == 1:
-                        continue
-                    n_idx = neighbor.GetIdx() - 1
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
-            else:
-                for neighbor in mol.Mol.GetAtomWithIdx(atom_idx).GetNeighbors():
-                    if neighbor.GetAtomicNum() == 1:
-                        continue
-                    n_idx = neighbor.GetIdx()
-                    if (n_idx not in atom_env[r - 1] and n_idx not in tmp):
-                        tmp.append(n_idx)
+        if mode == 'spherical':
+            neighbors = kd_tree.query_ball_point(mol.coords[idx], r=r*2)
+            for neighbor in neighbors:
+                atom = mol.OBMol.GetAtom(neighbor + 1)
```

Comment:
`neighbor` is `Atom` object. `Atom` has both `atomicnum`  and `idx0` property. Querying for OBAtom is redundant and toolkit dependent.